### PR TITLE
SONARJS-656 Support object spread/rest properties

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ObjectBindingPatternTree.java [hunk 1]
```diff
@@ -34,6 +34,7 @@
  * </pre>
  */
 @Beta
+// fixme do we need "extends DeclarationTree"?
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ObjectBindingPatternTree.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/expression/SpreadElementTree.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import com.google.common.annotations.Beta;
+import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
+
+/**
+ * Common interface for all types of Spread Element
+ * <pre>
+ *  ... {@link #element()}
+ * </pre>
+ *
+ * // todo where used
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/expression/SpreadElementTree.java]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
#### Analysis summary
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* INFO 1 info

Watch the comments in this conversation to review them.

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java [hunk 3]
```diff
@@ -356,6 +357,10 @@ private void buildExpression(Tree tree) {
       RestElementTree restElement = (RestElementTree) tree;
       buildExpression(restElement.element());
 
+    }  else if (tree.is(Kind.SPREAD_ELEMENT)) {
```
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @vilchik-elena do we still need the code for REST_ELEMENT on the lines above?

File: javascript-frontend/src/test/java/org/sonar/javascript/tree/impl/expression/SpreadElementTreeModelTest.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import static org.fest.assertions.Assertions.assertThat;
+
+public class SpreadElementTreeModelTest extends JavaScriptTreeModelTest {
+
+  @Test
+  public void spread_element() throws Exception {
+    SpreadElementTree tree = parse("var x = [1, ... expression];", Kind.SPREAD_ELEMENT);
+
+    assertThat(tree.is(Kind.SPREAD_ELEMENT)).isTrue();
+    assertThat(tree.ellipsis().text()).isEqualTo(JavaScriptPunctuator.ELLIPSIS.getValue());
```
Comment (pynicolas) [inline javascript-frontend/src/test/java/org/sonar/javascript/tree/impl/expression/SpreadElementTreeModelTest.java]: @vilchik-elena Shouldn't we also test `tree.element()`?

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java [hunk 5]
```diff
@@ -1446,26 +1450,29 @@ public InitializedBindingElementTreeImpl INITIALISER_NO_IN() {
...
+        b.zeroOrMore(f.newTuple53(b.token(JavaScriptPunctuator.COMMA), BINDING_PROPERTY())),
+        b.optional(b.token(JavaScriptPunctuator.COMMA))));
   }
 
   public BindingElementTree BINDING_PROPERTY() {
     return b.<BindingElementTree>nonterminal()
-      .is(
-        b.firstOf(
+      .is(b.firstOf(
+          REST_OBJECT_BINDING_ELEMENT(),
```
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java]: @vilchik-elena It seems to me that the rest should only be allowed as the end of the list: https://github.com/sebmarkbage/ecmascript-rest-spread/blob/master/Spec.md#rest-properties

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ObjectBindingPatternTree.java (hunk 1 above)
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ObjectBindingPatternTree.java]: @vilchik-elena I don't see why this should extend `DeclarationTree`.

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java (hunk 3 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @pynicolas looks like we don't need it as REST_ELEMENT is not expression anymore (it's replaced by SPREAD_ELEMENT). I will remove the previous clause

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java (hunk 5 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java]: @pynicolas Yep! but transpiler takes it. Try babel

Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java]: @vilchik-elena Could it be a bug in babel?

Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java]: @pynicolas Everything could happen:) I've created a question http://stackoverflow.com/questions/36429161/should-rest-property-of-object-binding-pattern-be-the-last

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ObjectBindingPatternTree.java (hunk 1 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ObjectBindingPatternTree.java]: @pynicolas fixed c567f7b00513895b4a232ed110034834495369db

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/expression/SpreadElementTree.java (hunk 2 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/expression/SpreadElementTree.java]: fixed c567f7b00513895b4a232ed110034834495369db

File: javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java (hunk 3 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/cfg/ControlFlowGraphBuilder.java]: @pynicolas fixed c567f7b00513895b4a232ed110034834495369db

File: javascript-frontend/src/test/java/org/sonar/javascript/tree/impl/expression/SpreadElementTreeModelTest.java (hunk 4 above)
Comment (vilchik-elena) [inline javascript-frontend/src/test/java/org/sonar/javascript/tree/impl/expression/SpreadElementTreeModelTest.java]: @pynicolas fixed c567f7b00513895b4a232ed110034834495369db

File: javascript-frontend/src/test/java/org/sonar/javascript/tree/impl/expression/SpreadElementTreeModelTest.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    assertThat(tree.element().is(Kind.IDENTIFIER_REFERENCE));
```
Comment (SonarTech) [inline javascript-frontend/src/test/java/org/sonar/javascript/tree/impl/expression/SpreadElementTreeModelTest.java]: CRITICAL Complete the assertion. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2970)

**Comment by SonarTech** [comment]
#### Analysis summary
SonarQube analysis reported 1 issue:
* CRITICAL 1 critical

Watch the comments in this conversation to review them.

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java (hunk 5 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java]: @pynicolas We have an answer, so i will fix it in grammar

**Comment by SonarTech** [comment]
#### Analysis summary
SonarQube analysis reported 1 issue

#### 1 unreported issues

Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:

1. [Tree.java#L115](https://github.com/SonarSource/sonar-javascript/blob/37c90517a3797ce5ca6abcfda657964702ba989c/javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/Tree.java#L115): MINOR "AstNodeType" is already extended by "GrammarRuleKey"; there is no need to implement it here. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1939)

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java [hunk 7]
```diff
@@ -1430,17 +1430,42 @@ public RestElementTreeImpl restObjectBindingElement(InternalSyntaxToken ellipsis
...
+      elements,
+      rCurlyBrace);
   }
 
-  public ObjectBindingPatternTreeImpl objectBindingPattern(InternalSyntaxToken lCurlyBrace, Optional<SeparatedList<Tree>> elements, InternalSyntaxToken rCurlyBrace) {
+  public ObjectBindingPatternTreeImpl objectBindingPattern2(InternalSyntaxToken lCurlyBrace, RestElementTree rest, InternalSyntaxToken rCurlyBrace) {
     return new ObjectBindingPatternTreeImpl(
       lCurlyBrace,
-      elements.isPresent() ? elements.get() : new SeparatedList<>(Collections.<Tree>emptyList(), Collections.<InternalSyntaxToken>emptyList()),
+      new SeparatedList<>(ImmutableList.<Tree>of(rest), Collections.<InternalSyntaxToken>emptyList()),
```
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: @vilchik-elena Why using both `ImmutableList.of()` and `Collections.emptyList()`?

Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: @pynicolas good question:)
`ImmutableList.<Tree>of(rest), ImmutableList.<InternalSyntaxToken>of()` is better for you?

Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: @vilchik-elena Yes, it's better. Both methods have same purpose.

Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: @pynicolas fixed in c8aceafffb5509a9a3f8af8e759911169970dc54

**Pull request merged**